### PR TITLE
feat(channel): add DingTalk knowledge base tools

### PR DESCRIPTION
**Pull request by iluv7**
## AgentScope Version

2.0.7.post1

## Description

Related issue: [#2247](https://github.com/agentscope-ai/agentscope/issues/2247#event-29834931529)

This PR is a follow-up to [#2285](https://github.com/agentscope-ai/agentscope/pull/2285#issuecomment-5392382645), which added `DingTalkChannel` but deferred DingTalk knowledge-base access because the original PR was already substantial. This PR completes that remaining part of the issue with a minimal, read-only integration that lets an Agent browse and read knowledge-base content visible to the current DingTalk user.

### What changed

- Added three read-only, Agent-callable DingTalk tools:
  - `ListKnowledgeBases` lists knowledge bases visible to the current DingTalk sender.
  - `ListKnowledgeNodes` browses the direct children of a knowledge-base root or folder and supports pagination.
  - `ReadKnowledgeDocument` reads a bounded range of an `ALIDOC` document and converts supported blocks to Markdown.
- Added DingTalk OpenAPI support for:
  - resolving the trusted channel sender's staff ID to a `unionId`;
  - listing user-visible Wiki workspaces;
  - listing and reading Wiki nodes;
  - reading document blocks with bounded ranges.
- Propagated the trusted `channel_user_id` through the channel wake-up path so knowledge tools are bound to the user who sent the current DingTalk message. The tools are not exposed to unrelated or background runs that do not have a trusted channel sender.
- Kept the new tools read-only and allowed by default. DingTalk remains responsible for authorization, and each request uses the current sender's `unionId` as `operatorId`.
- Added bounded document rendering for headings, paragraphs, ordered and unordered lists, and blockquotes. Unsupported rich blocks are represented by placeholders and reported to the Agent instead of being silently discarded.
- Added unit, service, gateway, dispatcher, and manual live E2E coverage.

### How to test

Run the related unit and service regression suite:

```bash
pytest -q \
  tests/channel_dingtalk_test.py \
  tests/channel_gateway_test.py \
  tests/service_chat_channel_delivery_test.py \
  tests/service_wakeup_dispatcher_test.py
```

Result:

```text
83 passed, 3 subtests passed
```

Run the checks for the changed files:

```bash
pre-commit run --files <changed files>
```

The Python AST, private-key detection, whitespace, trailing-comma, mypy, Black, flake8, pylint, and Pyroma checks passed.

For a live DingTalk regression, configure a real internal application with the required published contact, Wiki-node, workspace, and document read permissions, then run:

```bash
export DINGTALK_CLIENT_ID='...'
export DINGTALK_CLIENT_SECRET='...'
python tests/channel_dingtalk_knowledge_e2e.py
```

The manual runner verified the following real path:

```text
Stream inbound message
-> resolve the current sender
-> list visible knowledge bases
-> browse document nodes
-> read the target document
-> find the expected marker
-> reply with PASS in DingTalk
```

Observed result:

```text
PASS: knowledge base visible: AgentScope E2E KB
PASS: document node found: AgentScope E2E Runbook
PASS: document marker found: DINGTALK_KB_E2E_PASS_20260830
```

#### DingTalk Channel connected

5285c07b4f561f2397dd90c35db5c5db

#### Natural-language knowledge-base retrieval

8be50d7c7d49f73de74e00323a90f2c3

## Checklist

Please check the following items before code is ready to be reviewed.

- [x] An issue has been created for this PR
- [x] I have read the [CONTRIBUTING.md](https://github.com/agentscope-ai/agentscope/blob/main/CONTRIBUTING.md)
- [x] Docstrings are in Google style
- [ ] Related documentation has been updated (e.g. links, examples, etc.) in [documentation repository](https://github.com/agentscope-ai/docs)
- [x] Code is ready for review

### CODE REVIEW

**Comment by iluv7** [comment]
While testing the DingTalk Knowledge Base feature, I discovered a Web UI layout issue where the channel toggle switch extended beyond the card's boundaries; I went ahead and fixed it in this PR: #2466.

**Review by DavdGao — COMMENTED** [review]
Thanks for the contribution! This PR highlights an important design issue: since one channel serves many different channel users and chat groups, the chat service should be accessible via `channel_user_id` for tool equipment.

However, I'd suggest moving the channel user ID into `SessionRecord` as `SessionRecord.source_channel_user_id`, so that the SDK layer remains unaware of `channel_user_id`. This way, we only need to add an optional `channel_user_id` parameter to the existing `list_tools` method, rather than introducing a new abstract method.

**Comment by iluv7** [comment]
@DavdGao 
Thanks, that makes sense. I agree that the channel user identity belongs in SessionRecord rather than `AgentState.tool_context`, and extending the existing list_tools method is cleaner than introducing `list_tools_for_user`.
One detail I'd like to confirm: under `PER_CHAT`, multiple users in the same group share one session, while knowledge-base permissions are user-scoped. Should `source_channel_user_id` be refreshed for every inbound turn, or should user-scoped channel tools require `PER_CHAT_USER` to avoid one group member accessing tools under another member's identity?

**Comment by iluv7** [comment]
I’ve updated the implementation to keep the channel-specific identity in the application session layer.
The changes are:
- Added SessionRecord.source_channel_user_id.
- The ChannelGateway stores the trusted platform sender when creating a channel session and refreshes it for each new inbound turn.
- Redis and SQL storage now persist this field on session creation and update it when a new value is provided.
- Added an optional channel_user_id parameter to the existing ChannelBase.list_tools method.
- ChatService now reads the user ID from SessionRecord and passes it to list_tools when equipping channel tools.
- DingTalk uses this value to bind the knowledge-base tools to the corresponding platform user, while keeping it out of the model-visible tool arguments.
- Removed the additional list_tools_for_user method.
- Removed channel_user_id from AgentState.tool_context, run-trigger payloads, WakeupDispatcher, and the ChatService.run interface.

I also updated the Feishu list_tools override to accept the new optional parameter for interface compatibility only; its behavior remains unchanged.
The related channel, storage, and service tests have been updated, and all affected unit tests and pre-commit checks pass. Thanks again for pointing out the cleaner design.

**Comment by iluv7** [comment]
There is one remaining edge case with PER_CHAT sessions in group chats.
Normally, when the bot is idle and Bob starts a new turn, source_channel_user_id is updated to Bob, so the knowledge tools use Bob’s DingTalk unionId and only access resources available to Bob.
However, if Alice has already started a run and Bob mentions the bot while that run is still active, Bob’s message is injected into Alice’s run as a HintBlock. The toolkit was already equipped using Alice’s source_channel_user_id, so a knowledge-base request introduced by Bob’s hint could still execute with Alice’s permissions and return the result to the group.
Configuring the channel binding with PER_CHAT_USER avoids this issue because each (chat_id, channel_user_id) gets an independent session, run lock, context, and tool identity.
I don’t think we should change the global default from PER_CHAT to PER_CHAT_USER in this PR, since that would also change the group-chat behavior of other channels. For DingTalk channels with user-scoped knowledge-base tools, though, PER_CHAT_USER should probably be recommended or enforced. What do you think?

**Comment by DavdGao** [comment]
> There is one remaining edge case with PER_CHAT sessions in group chats. Normally, when the bot is idle and Bob starts a new turn, source_channel_user_id is updated to Bob, so the knowledge tools use Bob’s DingTalk unionId and only access resources available to Bob. However, if Alice has already started a run and Bob mentions the bot while that run is still active, Bob’s message is injected into Alice’s run as a HintBlock. The toolkit was already equipped using Alice’s source_channel_user_id, so a knowledge-base request introduced by Bob’s hint could still execute with Alice’s permissions and return the result to the group. Configuring the channel binding with PER_CHAT_USER avoids this issue because each (chat_id, channel_user_id) gets an independent session, run lock, context, and tool identity. I don’t think we should change the global default from PER_CHAT to PER_CHAT_USER in this PR, since that would also change the group-chat behavior of other channels. For DingTalk channels with user-scoped knowledge-base tools, though, PER_CHAT_USER should probably be recommended or enforced. What do you think?

@iluv7 Great catch. I don't have a solution yet, but this is a security concern that needs to be addressed before merging. We'll need to take some time to figure out the best way to handle it.

**Comment by iluv7** [comment]
@DavdGao Agreed, we should resolve this before merging. Thanks for looking into it! I’m happy to help work through the options and add multi-user group-chat tests once we settle on an approach.

**Comment by DavdGao** [comment]
> @DavdGao Agreed, we should resolve this before merging. Thanks for looking into it! I’m happy to help work through the options and add multi-user group-chat tests once we settle on an approach.

@iluv7  Yep, but the knowledge part itself is a good PR — so how about only enabling these knowledge tools in private chats for now? i.e. a check in chat_service where the tools are assembled.

**Comment by iluv7** [comment]
> > @DavdGao Agreed, we should resolve this before merging. Thanks for looking into it! I’m happy to help work through the options and add multi-user group-chat tests once we settle on an approach.
> 
> @iluv7 Yep, but the knowledge part itself is a good PR — so how about only enabling these knowledge tools in private chats for now? i.e. a check in chat_service where the tools are assembled.

Sounds good. I’ll restrict the DingTalk knowledge tools to private chats for now and add tests to ensure they aren’t exposed in group chats.

**Comment by iluv7** [comment]
@DavdGao 
Done — it was a pretty small change. DingTalk knowledge tools are now only available in private chats, while the other DingTalk tools still work in both private and group chats. Other channels and the existing PER_CHAT behavior are unchanged. I also added tests for both cases.

**Comment by iluv7** [comment]
Resolved the conflicts with the latest `main` and updated the implementation to work with the new `SessionOrigin` / `ChannelOrigin` structure. The existing `source_channel_user_id` behavior and private-chat-only restriction for DingTalk knowledge tools are preserved. Related tests are passing.

**Comment by iluv7** [comment]
The unit-test failure in `MCPRuntimeHeadersTest.test_owning_the_client_keeps_transport_defaults` is unrelated to the DingTalk changes in this PR. MCP 1.30.0 was published after the previous successful CI runs and intentionally changed the default HTTP client redirect behavior from `follow_redirects=True` to `False`. Because AgentScope currently depends on `mcp<2.0.0` and CI installs dependencies without a lock file, this run picked up MCP 1.30.0 while the test still hard-coded the previous default.

I opened #2542 to make the test derive the expected values from MCP's own HTTP client factory. The fix has been verified against both MCP 1.28.1 and 1.30.0. Once #2542 is merged and this branch is updated with the latest `main`, this unrelated failure should be resolved.

**Comment by DavdGao** [comment]
Pushed a few changes on top of your branch, main points:

- Wiki browsing is now declared once on `ChannelBase` (`list_wiki_spaces` / `list_wiki_nodes` / `read_wiki_document`) behind `ChannelCapability.wiki`, and the three tools moved to `channel/_tools` so any channel with a wiki gets them. DingTalk just implements the contract now. Feishu and WeCom model wikis the same way, so they only need to fill in the three methods later.
- Named it wiki instead of knowledge base, since this repo already uses knowledge base for the RAG ones an agent can be given.
- `ChannelOrigin.channel_user_id` instead of a new top-level `SessionRecord` field, and the private-chat rule moved into the chat service so it is not DingTalk-specific.
- Table blocks are rendered as real Markdown tables now. Per the block docs a table block carries its cell text in `cells`, and the row count field is spelled `rolSize` — the old `rowSize` never matched anything, so tables were always coming back as a placeholder with `? rows`. Also dropped `unsupported_block_types`, the inline placeholder already says the type at the position it stood in.

One thing I could not confirm from the docs: `tableRow` and `tableCell` are listed as block types of their own, but a table block already carries `cells`. Do they also show up flattened in the top-level block list? If they do, the cell text would be duplicated and they would count against `startIndex`/`endIndex` paging. You have a live environment — could you check what a document with a table actually returns?

File: src/agentscope/app/channel/_dingtalk/_channel.py [hunk 1]
```diff
@@ -1104,3 +1235,117 @@ def _media_description(
...
+    for block in sorted(blocks, key=_block_index):
+        rendered = _render_block(block)
+        added = len(rendered) + (2 if parts else 0)
+        if total + added > _MAX_DOCUMENT_CHARS:
+            # A single oversized block would otherwise return nothing and
+            # no way to resume, so emit a prefix and stop before it.
+            if not parts:
+                parts.append(rendered[:_MAX_DOCUMENT_CHARS])
+                last_index = _block_index(block)
+            return "\n\n".join(parts), last_index + 1
```
[the remark this answers is not in the document]
Comment (DavdGao) [reply src/agentscope/app/channel/_dingtalk/_channel.py]: Good catch, this was losing content. Fixed by never splitting a block: if the first one is over budget it now comes back whole, so the resume index always points at a block nobody has seen.

File: src/agentscope/app/channel/_dingtalk/_channel.py [hunk 2]
```diff
@@ -1104,3 +1235,117 @@ def _media_description(
...
+
+    DingTalk carries the cell text inside the block as a 2-D array, so a
+    table reaches the agent whole rather than as a placeholder. Note the
+    row count is spelled ``rolSize`` by the platform.
+    """
+    rows = [
+        [str(cell) for cell in row]
+        for row in detail.get("cells") or []
+        if isinstance(row, list)
+    ]
```
[the remark this answers is not in the document]
Comment (DavdGao) [reply src/agentscope/app/channel/_dingtalk/_channel.py:1336]: Right, fixed. Cell text now escapes its own backslashes and pipes and flattens line breaks before the join.

File: tests/channel_dingtalk_test.py [hunk 3]
```diff
@@ -1464,6 +1764,91 @@ async def test_search_users_resolves_profile_details(self) -> None:
...
+
+        spaces = await api.list_wiki_workspaces("staff-1", 20)
+        nodes = await api.list_wiki_nodes("staff-1", "root-1", 50, "node-next")
+        node = await api.get_wiki_node("staff-1", "doc/1")
+        blocks = await api.read_document_blocks("staff-1", "doc/1", 0, 49)
+
+        self.assertEqual(spaces["nextToken"], "space-next")
+        self.assertEqual(nodes["nodes"][0]["nodeId"], "doc/1")
+        self.assertEqual(node["nodeId"], "doc/1")
+        self.assertEqual(blocks[0]["paragraph"]["text"], "hello")
```
[the remark this answers is not in the document]
Comment (DavdGao) [reply tests/channel_dingtalk_test.py]: Agreed, this one is in the repo test rules too. Rewrote it to assert the four responses and the full list of requests as whole structures.

File: tests/channel_dingtalk_test.py [hunk 4]
```diff
@@ -1464,6 +1764,91 @@ async def test_search_users_resolves_profile_details(self) -> None:
...
+                        "maxResults": 20,
+                        "withPermissionRole": True,
+                    },
+                },
+            ),
+        )
+        self.assertEqual(http.gets[1][1]["params"]["parentNodeId"], "root-1")
+        self.assertEqual(http.gets[1][1]["params"]["nextToken"], "node-next")
+        self.assertTrue(http.gets[2][0].endswith("/nodes/doc%2F1"))
+        self.assertTrue(http.gets[3][0].endswith("/documents/doc%2F1/blocks"))
```
[the remark this answers is not in the document]
Comment (DavdGao) [reply tests/channel_dingtalk_test.py]: Same fix as the comment above — http.gets is now asserted as one complete list, so an endpoint or param change shows up.

File: src/agentscope/app/channel/_tools/_list_wiki_spaces.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        default="",
+        max_length=2048,
+        description="Pagination token returned by a previous call.",
+    )
+
+
+class ListWikiSpaces(_WikiToolBase):
+    """List wiki spaces readable by the current sender."""
+
+    name: str = "ListWikiSpaces"
```
[the remark this answers is not in the document]
Comment (DavdGao) [reply src/agentscope/app/channel/_tools/_list_wiki_spaces.py:30]: The Wiki naming is intentional. This repo already uses knowledge base for the RAG ones an agent can be given, and an agent holding both sets of tools would have to disambiguate them. Both DingTalk and Feishu call these APIs wiki anyway. @iluv7 could you update the PR description to match?

File: src/agentscope/app/channel/_gateway.py [hunk 6]
```diff
@@ -324,6 +324,7 @@ async def _ensure_session(
                 channel_id=record.id,
                 chat_id=event.chat_id,
                 chat_name=event.chat_name or None,
+                channel_user_id=event.channel_user_id or None,
```
[the remark this answers is not in the document]
Comment (DavdGao) [reply src/agentscope/app/channel/_gateway.py:327]: Correct that a session created before this field existed keeps None and so never gets the wiki tools. Not fixing it here though: origin is documented as fixed at creation and never rewritten, and upsert_session ignores origin on the update path, so a backfill means making origin mutable for every caller. That is a bigger change than this PR should carry. Starting a new session picks the field up, and the field docstring says so. Happy to do it separately if we want the migration.

**Comment by iluv7** [comment]
> Pushed a few changes on top of your branch, main points:
> 
> * Wiki browsing is now declared once on `ChannelBase` (`list_wiki_spaces` / `list_wiki_nodes` / `read_wiki_document`) behind `ChannelCapability.wiki`, and the three tools moved to `channel/_tools` so any channel with a wiki gets them. DingTalk just implements the contract now. Feishu and WeCom model wikis the same way, so they only need to fill in the three methods later.
> * Named it wiki instead of knowledge base, since this repo already uses knowledge base for the RAG ones an agent can be given.
> * `ChannelOrigin.channel_user_id` instead of a new top-level `SessionRecord` field, and the private-chat rule moved into the chat service so it is not DingTalk-specific.
> * Table blocks are rendered as real Markdown tables now. Per the block docs a table block carries its cell text in `cells`, and the row count field is spelled `rolSize` — the old `rowSize` never matched anything, so tables were always coming back as a placeholder with `? rows`. Also dropped `unsupported_block_types`, the inline placeholder already says the type at the position it stood in.
> 
> One thing I could not confirm from the docs: `tableRow` and `tableCell` are listed as block types of their own, but a table block already carries `cells`. Do they also show up flattened in the top-level block list? If they do, the cell text would be duplicated and they would count against `startIndex`/`endIndex` paging. You have a live environment — could you check what a document with a table actually returns?


Okay, I'll check that now.

**Comment by iluv7** [comment]
I created a document named `AgentScope Table Blocks E2E` in the test Wiki and added a 3×3 table with unique content in every cell. The test connected through DingTalk Stream, received a private message to identify the current user, located the document through the Wiki APIs, and then called the document blocks API with different `startIndex` / `endIndex` values.

The full response contained two top-level blocks:

```text
0: table
1: paragraph
```

The table block contained all nine values under `table.cells`. No top-level `tableRow` or `tableCell` blocks were returned.

I also checked pagination one index at a time:

```text
0..0: table
1..1: paragraph
2..2: empty
```

So the whole table counts as a single block. There is no duplicated cell content, and `tableRow` / `tableCell` do not affect `startIndex` / `endIndex` pagination in this live response.

One additional finding: the live API returned `rowSize`, not `rolSize`:

```json
{
  "table": {
    "cells": [
      ["R1C1UNIQUE", "R1C2UNIQUE", "R1C3UNIQUE"],
      ["R2C1UNIQUE", "R2C2UNIQUE", "R2C3UNIQUE"],
      ["R3C1UNIQUE", "R3C2UNIQUE", "R3C3UNIQUE"]
    ],
    "colSize": 3,
    "rowSize": 3
  }
}
```

Rendering works when `cells` is present, but I think the empty-table fallback and the corresponding test fixture should use `rowSize`.

**Comment by DavdGao** [comment]
Thanks for running that, both answers are exactly what was needed.

On `rowSize` you are right and I was wrong — reverted in the latest push, and the test fixture uses `rowSize` too now. For the record: the `rolSize` spelling I switched to comes from the insert-block schema, and the reference I took it from has no read-response example for a table at all. I extrapolated the write side onto the read path, which was not justified. Your live response is the only direct evidence about what the blocks API actually returns, and it agrees with what this PR had originally. So the only real defect there was dropping `cells`, which is what the Markdown table rendering now fixes.

On `tableRow` / `tableCell`, that settles it: the table is one block, its cells are not flattened into the top-level list, and it costs one index in `startIndex` / `endIndex`. No change needed, and the paging stays correct.

**Review by DavdGao — APPROVED** [review]
